File: packages/typescript/src/features/plugin-manager/services/pluginRegistryService.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { logger } from "../../../logger.ts";
import type { PluginMetadata } from "../types.ts";

// ---------------------------------------------------------------------------
// Registry URLs — next branch
// ---------------------------------------------------------------------------

const GENERATED_REGISTRY_URL =
	"https://raw.githubusercontent.com/elizaos-plugins/registry/next/generated-registry.json";
const INDEX_REGISTRY_URL =
	"https://raw.githubusercontent.com/elizaos-plugins/registry/next/index.json";

const CACHE_DURATION = 3_600_000; // 1 hour

// ---------------------------------------------------------------------------
// Local plugins directory (relative to cwd)
// ---------------------------------------------------------------------------

const LOCAL_PLUGINS_DIR = "plugins";

// ---------------------------------------------------------------------------
// Wire types for the generated-registry.json format
// ---------------------------------------------------------------------------

interface GeneratedRegistryEntry {
	git: {
		repo: string;
		v0: { version: string | null; branch: string | null };
		v1: { version: string | null; branch: string | null };
		v2: { version: string | null; branch: string | null };
	};
	npm: {
		repo: string;
		v0: string | null;
		v1: string | null;
		v2: string | null;
		v0CoreRange: string | null;
		v1CoreRange: string | null;
		v2CoreRange: string | null;
	};
	supports: { v0: boolean; v1: boolean; v2: boolean };
	description: string;
	homepage: string | null;
	topics: string[];
	stargazers_count: number;
	language: string;
}

interface GeneratedRegistryFile {
	lastUpdatedAt: string;
	registry: Record<string, GeneratedRegistryEntry>;
	apps?: Record<string, GeneratedRegistryEntry>;
}

// ---------------------------------------------------------------------------
// Normalised plugin representation
// ---------------------------------------------------------------------------

export interface RegistryPlugin {
	name: string;
	gitRepo: string;
	gitUrl: string;
	description: string;
	homepage: string | null;
	topics: string[];
	stars: number;
	language: string;
	npm: {
		package: string;
		v0Version: string | null;
		v1Version: string | null;
		v2Version: string | null;
		v0CoreRange: string | null;
		v1CoreRange: string | null;
		v2CoreRange: string | null;
	};
	git: {
		v0Branch: string | null;
		v1Branch: string | null;
		v2Branch: string | null;
	};
	supports: { v0: boolean; v1: boolean; v2: boolean };
	// App/Viewer extensions
	viewer?: {
		url: string;
		embedParams?: Record<string, string>;
		postMessageAuth?: boolean;
		sandbox?: string;
	};
	launchType?: "connect" | "local";
	launchUrl?: string;
	displayName?: string;
	kind?: string;
	// App-specific metadata
	category?: string;
	capabilities?: string[];
	icon?: string | null;
}

export interface PluginSearchResult {
	name: string;
	description: string;
	score: number;
	tags: string[];
	version: string | null;
	npmPackage: string;
	repository: string;
	stars: number;
	supports: { v0: boolean; v1: boolean; v2: boolean };
}

export interface CloneResult {
	success: boolean;
	error?: string;
	pluginName?: string;
	localPath?: string;
	hasTests?: boolean;
	dependencies?: Record<string, string>;
}

// ---------------------------------------------------------------------------
// In-memory cache
// ---------------------------------------------------------------------------

let registryCache: {
	plugins: Map<string, RegistryPlugin>;
	timestamp: number;
} | null = null;

export function resetRegistryCache(): void {
	registryCache = null;
}

// ---------------------------------------------------------------------------
// Fetching & parsing
// ---------------------------------------------------------------------------

function entryToPlugin(
	name: string,
	e: GeneratedRegistryEntry & { kind?: string },
): RegistryPlugin {
	return {
		name,
		gitRepo: e.git.repo,
		gitUrl: `https://github.com/${e.git.repo}.git`,
		description: e.description || "",
		homepage: e.homepage,
		topics: e.topics || [],
		stars: e.stargazers_count || 0,
		language: e.language || "TypeScript",
		npm: {
			package: e.npm.repo,
			v0Version: e.npm.v0,
			v1Version: e.npm.v1,
			v2Version: e.npm.v2,
			v0CoreRange: e.npm.v0CoreRange,
			v1CoreRange: e.npm.v1CoreRange,
			v2CoreRange: e.npm.v2CoreRange,
		},
		git: {
			v0Branch: e.git.v0?.branch ?? null,
			v1Branch: e.git.v1?.branch ?? null,
			v2Branch: e.git.v2?.branch ?? null,
		},
		supports: e.supports,
		kind: e.kind,
	};
}

function stubPlugin(name: string, gitRef: string): RegistryPlugin {
	const repo = gitRef.replace(/^github:/, "");
	return {
		name,
		gitRepo: repo,
		gitUrl: `https://github.com/${repo}.git`,
		description: "",
		homepage: null,
		topics: [],
		stars: 0,
		language: "TypeScript",
		npm: {
			package: name,
			v0Version: null,
			v1Version: null,
			v2Version: null,
			v0CoreRange: null,
			v1CoreRange: null,
			v2CoreRange: null,
		},
		git: { v0Branch: null, v1Branch: null, v2Branch: "next" },
		supports: { v0: false, v1: false, v2: false },
	};
}

// ---------------------------------------------------------------------------
// Local plugin discovery - scans plugins/ for elizaos.plugin.json files
// ---------------------------------------------------------------------------

interface LocalPluginJson {
	id?: string;
	name?: string;
	description?: string;
	version?: string;
	kind?: string;
	app?: {
		displayName?: string;
		category?: string;
		launchType?: "connect" | "local";
		launchUrl?: string;
		capabilities?: string[];
	};
	viewer?: {
		url: string;
		embedParams?: Record<string, string>;
		postMessageAuth?: boolean;
		sandbox?: string;
	};
	configSchema?: Record<string, unknown>;
	keywords?: string[];
	author?: string;
	homepage?: string;
	repository?: string;
}

function localPluginToRegistry(
	pluginJson: LocalPluginJson,
	dirName: string,
): RegistryPlugin {
	const name = pluginJson.id || `@elizaos/${dirName}`;
	const displayName = pluginJson.app?.displayName || pluginJson.name || dirName;
	const description = pluginJson.description || "";
	const homepage = pluginJson.homepage || null;
	const keywords = pluginJson.keywords || [];
	const repo =
		pluginJson.repository
			?.replace("https://github.com/", "")
			.replace(".git", "") || `elizaos/${dirName}`;

	return {
		name,
		gitRepo: repo,
		gitUrl: pluginJson.repository || `https://github.com/${repo}.git`,
		description,
		displayName,
		homepage,
		topics: keywords,
		stars: 0,
		language: "TypeScript",
		npm: {
			package: name,
			v0Version: null,
			v1Version: pluginJson.version || null,
			v2Version: pluginJson.version || null,
			v0CoreRange: null,
			v1CoreRange: null,
			v2CoreRange: null,
		},
		git: { v0Branch: null, v1Branch: null, v2Branch: "main" },
		supports: { v0: false, v1: true, v2: true },
		kind: pluginJson.kind,
		launchType: pluginJson.app?.launchType,
		launchUrl: pluginJson.app?.launchUrl,
		viewer: pluginJson.viewer,
		// App-specific metadata
		category: pluginJson.app?.category,
		capabilities: pluginJson.app?.capabilities || [],
		icon: null,
	};
}

async function scanLocalPlugins(): Promise<Map<string, RegistryPlugin>> {
	const plugins = new Map<string, RegistryPlugin>();
	const pluginsDir = path.resolve(process.cwd(), LOCAL_PLUGINS_DIR);

	if (!fs.existsSync(pluginsDir)) {
		return plugins;
	}

	try {
		const entries = fs.readdirSync(pluginsDir, { withFileTypes: true });

		for (const entry of entries) {
			if (!entry.isDirectory()) continue;

			const pluginJsonPath = path.join(
				pluginsDir,
				entry.name,
				"elizaos.plugin.json",
			);
			if (!fs.existsSync(pluginJsonPath)) continue;

			try {
				const content = fs.readFileSync(pluginJsonPath, "utf-8");
				const pluginJson = JSON.parse(content) as LocalPluginJson;
				const plugin = localPluginToRegistry(pluginJson, entry.name);
				plugins.set(plugin.name, plugin);
				logger.debug(
					`[registry] Found local plugin: ${plugin.name} (${entry.name})`,
				);
			} catch (err) {
				logger.warn(
					`[registry] Failed to parse ${pluginJsonPath}: ${errMsg(err)}`,
				);
			}
		}

		if (plugins.size > 0) {
			logger.info(
				`[registry] Loaded ${plugins.size} local plugins from ${pluginsDir}`,
			);
		}
	} catch (err) {
		logger.warn(`[registry] Failed to scan local plugins: ${errMsg(err)}`);
	}

	return plugins;
}

async function fetchGeneratedRegistry(): Promise<Map<string, RegistryPlugin>> {
	const response = await fetch(GENERATED_REGISTRY_URL);
	if (!response.ok) {
		throw new Error(
			`generated-registry.json: ${response.status} ${response.statusText}`,
		);
	}
	const data = (await response.json()) as GeneratedRegistryFile;
	const plugins = new Map<string, RegistryPlugin>();
	for (const [name, entry] of Object.entries(data.registry)) {
		plugins.set(name, entryToPlugin(name, entry));
	}
	if (data.apps) {
		for (const [name, entry] of Object.entries(data.apps)) {
			plugins.set(name, entryToPlugin(name, { ...entry, kind: "app" }));
		}
	}
	return plugins;
}

async function fetchIndexRegistry(): Promise<Map<string, RegistryPlugin>> {
	const response = await fetch(INDEX_REGISTRY_URL);
	if (!response.ok) {
		throw new Error(`index.json: ${response.status} ${response.statusText}`);
	}
	const data = (await response.json()) as Record<string, string>;
	const plugins = new Map<string, RegistryPlugin>();
	for (const [name, gitRef] of Object.entries(data)) {
		plugins.set(name, stubPlugin(name, gitRef));
	}
	return plugins;
}

/**
 * Load the plugin registry from the next@registry branch.
 * Tries generated-registry.json first, falls back to index.json.
 * Also scans local plugins/ directory for elizaos.plugin.json files.
 * Local plugins override remote registry entries.
 * Cached in-memory for 1 hour.
 */
export async function loadRegistry(): Promise<Map<string, RegistryPlugin>> {
	if (registryCache && Date.now() - registryCache.timestamp < CACHE_DURATION) {
		return registryCache.plugins;
	}

	logger.info("[registry] Fetching from next@registry...");

	let plugins: Map<string, RegistryPlugin> = new Map();
	try {
		plugins = await fetchGeneratedRegistry();
		logger.info(
			`[registry] Loaded ${plugins.size} plugins (generated-registry.json)`,
		);
	} catch (err) {
		logger.warn(
			`[registry] generated-registry.json unavailable: ${errMsg(err)}, falling back to index.json`,
		);
		try {
			plugins = await fetchIndexRegistry();
			logger.info(`[registry] Loaded ${plugins.size} plugins (index.json)`);
		} catch (err2) {
			logger.warn(
				`[registry] index.json also unavailable: ${errMsg(err2)}, using local plugins only`,
			);
			// Continue with empty remote registry - local plugins will still be added
		}
	}

	// Merge local plugins (they override remote registry entries)
	const localPlugins = await scanLocalPlugins();
	for (const [name, plugin] of localPlugins) {
		plugins.set(name, plugin);
	}

	registryCache = { plugins, timestamp: Date.now() };
	return plugins;
}

// ---------------------------------------------------------------------------
// Lookup helpers
// ---------------------------------------------------------------------------

/** Resolve a plugin by name with fuzzy matching (exact -> @elizaos/ prefix -> bare name). */
function resolvePlugin(
	registry: Map<string, RegistryPlugin>,
	name: string,
): RegistryPlugin | null {
	let p = registry.get(name);
	if (p) return p;

	if (!name.startsWith("@")) {
		p = registry.get(`@elizaos/${name}`);
		if (p) return p;
	}

	const bare = name.replace(/^@[^/]+\//, "");
	for (const [key, value] of registry) {
		if (key.endsWith(`/${bare}`) || key === bare) return value;
	}

	return null;
}

export async function getRegistryEntry(
	name: string,
): Promise<RegistryPlugin | null> {
	return resolvePlugin(await loadRegistry(), name);
}

// ---------------------------------------------------------------------------
// RegistryPlugin -> PluginMetadata conversion
// ---------------------------------------------------------------------------

function toMetadata(p: RegistryPlugin): PluginMetadata {
	const author = p.gitRepo.split("/")[0] || "unknown";
	return {
		name: p.name,
		description: p.description,
		author,
		repository: `https://github.com/${p.gitRepo}`,
		versions: [p.npm.v0Version, p.npm.v1Version, p.npm.v2Version].filter(
			(v): v is string => v !== null,
		),
		latestVersion:
			p.npm.v2Version || p.npm.v1Version || p.npm.v0Version || "unknown",
		runtimeVersion: p.supports.v2 ? "v2" : p.supports.v1 ? "v1" : "v0",
		maintainer: author,
		tags: p.topics,
		categories: [],
	};
}

// ---------------------------------------------------------------------------
// Search
// ---------------------------------------------------------------------------

function computeSearchScore(plugin: RegistryPlugin, query: string): number {
	const lq = query.toLowerCase();
	const terms = lq.split(/\s+/).filter((t) => t.length > 1);
	const ln = plugin.name.toLowerCase();
	const ld = plugin.description.toLowerCase();

	let score = 0;

	if (ln === lq || ln === `@elizaos/${lq}`) score += 100;
	else if (ln.includes(lq)) score += 50;

	if (ld.includes(lq)) score += 30;
	for (const t of plugin.topics) if (t.toLowerCase().includes(lq)) score += 25;

	for (const term of terms) {
		if (ln.includes(term)) score += 15;
		if (ld.includes(term)) score += 10;
		for (const t of plugin.topics)
			if (t.toLowerCase().includes(term)) score += 8;
	}

	// Popularity bonus only when there's already a text match
	if (score > 0) {
		if (plugin.stars > 100) score += 5;
		if (plugin.stars > 500) score += 5;
		if (plugin.stars > 1000) score += 5;
	}

	return score;
}

export async function searchPluginsByContent(
	query: string,
	limit = 10,
): Promise<PluginSearchResult[]> {
	const registry = await loadRegistry();
	const scored: Array<{ plugin: RegistryPlugin; score: number }> = [];

	for (const plugin of registry.values()) {
		const score = computeSearchScore(plugin, query);
		if (score > 0) scored.push({ plugin, score });
	}

	scored.sort((a, b) => b.score - a.score || b.plugin.stars - a.plugin.stars);
	const maxScore = scored[0]?.score || 1;

	return scored.slice(0, limit).map(({ plugin, score }) => ({
		name: plugin.name,
		description: plugin.description,
		score: score / maxScore,
		tags: plugin.topics,
		version:
			plugin.npm.v2Version || plugin.npm.v1Version || plugin.npm.v0Version,
		npmPackage: plugin.npm.package,
		repository: `https://github.com/${plugin.gitRepo}`,
		stars: plugin.stars,
		supports: plugin.supports,
	}));
}

export async function getPluginDetails(
	name: string,
): Promise<PluginMetadata | null> {
	const plugin = resolvePlugin(await loadRegistry(), name);
	return plugin ? toMetadata(plugin) : null;
}

export async function getAllPlugins(): Promise<PluginMetadata[]> {
	const registry = await loadRegistry();
	return Array.from(registry.values(), toMetadata);
}

// ---------------------------------------------------------------------------
// Legacy support for non-app plugins
// ---------------------------------------------------------------------------

export async function listNonAppPlugins(): Promise<RegistryPlugin[]> {
	const registry = await loadRegistry();
	return Array.from(registry.values()).filter(
		(p) => p.kind !== "app" && !p.displayName,
	);
}

export async function searchNonAppPlugins(
	query: string,
	limit = 10,
): Promise<PluginSearchResult[]> {
	const registry = await loadRegistry();
	const scored: Array<{ plugin: RegistryPlugin; score: number }> = [];

	for (const plugin of registry.values()) {
		if (plugin.kind === "app" || plugin.displayName) continue;
		const score = computeSearchScore(plugin, query);
		if (score > 0) scored.push({ plugin, score });
	}

	scored.sort((a, b) => b.score - a.score || b.plugin.stars - a.plugin.stars);
	const maxScore = scored[0]?.score || 1;

	return scored.slice(0, limit).map(({ plugin, score }) => ({
		name: plugin.name,
		description: plugin.description,
		score: score / maxScore,
		tags: plugin.topics,
		version:
			plugin.npm.v2Version || plugin.npm.v1Version || plugin.npm.v0Version,
		npmPackage: plugin.npm.package,
		repository: `https://github.com/${plugin.gitRepo}`,
		stars: plugin.stars,
		supports: plugin.supports,
	}));
}

export async function refreshRegistry(): Promise<Map<string, RegistryPlugin>> {
	resetRegistryCache();
	return loadRegistry();
}

// ---------------------------------------------------------------------------
// Clone
// ---------------------------------------------------------------------------

export async function clonePlugin(pluginName: string): Promise<CloneResult> {
	logger.info(`[registry] Cloning plugin: ${pluginName}`);

	const plugin = resolvePlugin(await loadRegistry(), pluginName);
	if (!plugin) {
		return {
			success: false,
			error: `Plugin "${pluginName}" not found in registry`,
		};
	}

	const pathMod = await import("node:path");
	const fsMod = await import("node:fs/promises");
	const { promisify } = await import("node:util");
	const execAsync = promisify((await import("node:child_process")).exec);

	const cloneDir = pathMod.join(
		process.cwd(),
		"cloned-plugins",
		plugin.name.replace(/^@[^/]+\//, ""),
	);
	await fsMod.mkdir(cloneDir, { recursive: true });

	const branch = plugin.git.v2Branch || plugin.git.v1Branch || "next";
	await execAsync(
		`git clone --branch ${branch} --single-branch --depth 1 ${plugin.gitUrl} ${cloneDir}`,
	);

	let hasTests = false;
	let dependencies: Record<string, string> = {};
	try {
		const pkg = JSON.parse(
			await fsMod.readFile(pathMod.join(cloneDir, "package.json"), "utf-8"),
		) as {
			scripts?: Record<string, string>;
			devDependencies?: Record<string, string>;
			dependencies?: Record<string, string>;
		};
		hasTests = !!(pkg.scripts?.test || pkg.devDependencies?.vitest);
		dependencies = pkg.dependencies || {};
	} catch {
		logger.warn(`[registry] No package.json at repo root for ${plugin.name}`);
	}

	return {
		success: true,
		pluginName: plugin.name,
		localPath: cloneDir,
		hasTests,
		dependencies,
	};
}

// ---------------------------------------------------------------------------
// Util
// ---------------------------------------------------------------------------

function errMsg(err: unknown): string {
	return err instanceof Error ? err.message : String(err);
}

File: packages/python/elizaos/core_capabilities/plugin_manager/services/plugin_registry_service.py
"""Plugin registry service.

Fetches, caches, and searches plugin metadata from the elizaOS
remote registry (``generated-registry.json`` on the ``next`` branch).

Ported from plugin-manager/services/pluginRegistryService.ts.

The Python port adapts ``fetch`` to ``httpx`` (async) and ``fs``/``path``
to ``pathlib``.  The in-memory cache, fuzzy resolution, and search scoring
are preserved.
"""

from __future__ import annotations

import json
import logging
import re
import time
from pathlib import Path
from typing import TYPE_CHECKING, Any, ClassVar

from elizaos.types import Service

from ..types import PluginMetadata, PluginSearchResult, RegistryPlugin

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime

logger = logging.getLogger("elizaos.plugin_manager.registry")

# ---------------------------------------------------------------------------
# Registry URLs — next branch
# ---------------------------------------------------------------------------

_GENERATED_REGISTRY_URL = (
    "https://raw.githubusercontent.com/elizaos-plugins/registry/next/generated-registry.json"
)
_INDEX_REGISTRY_URL = "https://raw.githubusercontent.com/elizaos-plugins/registry/next/index.json"

_CACHE_DURATION = 3_600  # 1 hour in seconds

_LOCAL_PLUGINS_DIR = "plugins"

# ---------------------------------------------------------------------------
# Internal cache
# ---------------------------------------------------------------------------

_registry_cache: dict[str, Any] | None = None  # {"plugins": dict, "timestamp": float}


def reset_registry_cache() -> None:
    global _registry_cache
    _registry_cache = None


# ---------------------------------------------------------------------------
# Parsing helpers
# ---------------------------------------------------------------------------


def _entry_to_plugin(name: str, entry: dict) -> RegistryPlugin:
    git = entry.get("git", {})
    npm = entry.get("npm", {})
    supports = entry.get("supports", {})
    return RegistryPlugin(
        name=name,
        git_repo=git.get("repo", ""),
        git_url=f"https://github.com/{git.get('repo', '')}.git",
        description=entry.get("description", ""),
        homepage=entry.get("homepage"),
        topics=entry.get("topics", []),
        stars=entry.get("stargazers_count", 0),
        language=entry.get("language", "TypeScript"),
        npm_package=npm.get("repo", name),
        npm_v0_version=npm.get("v0"),
        npm_v1_version=npm.get("v1"),
        npm_v2_version=npm.get("v2"),
        supports_v0=supports.get("v0", False),
        supports_v1=supports.get("v1", False),
        supports_v2=supports.get("v2", False),
        kind=entry.get("kind"),
    )


def _stub_plugin(name: str, git_ref: str) -> RegistryPlugin:
    repo = git_ref.removeprefix("github:")
    return RegistryPlugin(
        name=name,
        git_repo=repo,
        git_url=f"https://github.com/{repo}.git",
        description="",
        homepage=None,
        topics=[],
        stars=0,
        language="TypeScript",
        npm_package=name,
        supports_v2=False,
    )


def _to_metadata(plugin: RegistryPlugin) -> PluginMetadata:
    author = plugin.git_repo.split("/")[0] if "/" in plugin.git_repo else "unknown"
    versions = [
        v
        for v in (plugin.npm_v0_version, plugin.npm_v1_version, plugin.npm_v2_version)
        if v is not None
    ]
    latest = plugin.npm_v2_version or plugin.npm_v1_version or plugin.npm_v0_version or "unknown"
    rv = "v2" if plugin.supports_v2 else ("v1" if plugin.supports_v1 else "v0")
    return PluginMetadata(
        name=plugin.name,
        description=plugin.description,
        author=author,
        repository=f"https://github.com/{plugin.git_repo}",
        versions=versions,
        latest_version=latest,
        runtime_version=rv,
        maintainer=author,
        tags=plugin.topics,
        categories=[],
    )


# ---------------------------------------------------------------------------
# Fetch
# ---------------------------------------------------------------------------


async def _fetch_json(url: str) -> Any:
    """Fetch JSON from *url* using httpx if available, falling back to urllib."""
    try:
        httpx = __import__("httpx")

        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(url)
            resp.raise_for_status()
            return resp.json()
    except ImportError:
        import urllib.request

        with urllib.request.urlopen(url, timeout=30) as resp:  # noqa: S310
            return json.loads(resp.read())


async def _fetch_generated_registry() -> dict[str, RegistryPlugin]:
    data = await _fetch_json(_GENERATED_REGISTRY_URL)
    plugins: dict[str, RegistryPlugin] = {}
    for name, entry in data.get("registry", {}).items():
        plugins[name] = _entry_to_plugin(name, entry)
    for name, entry in data.get("apps", {}).items():
        entry_with_kind = {**entry, "kind": "app"}
        plugins[name] = _entry_to_plugin(name, entry_with_kind)
    return plugins


async def _fetch_index_registry() -> dict[str, RegistryPlugin]:
    data = await _fetch_json(_INDEX_REGISTRY_URL)
    plugins: dict[str, RegistryPlugin] = {}
    for name, git_ref in data.items():
        plugins[name] = _stub_plugin(name, git_ref)
    return plugins


def _scan_local_plugins() -> dict[str, RegistryPlugin]:
    """Scan the local ``plugins/`` directory for ``elizaos.plugin.json`` files."""
    plugins: dict[str, RegistryPlugin] = {}
    plugins_dir = Path.cwd() / _LOCAL_PLUGINS_DIR
    if not plugins_dir.is_dir():
        return plugins

    for entry in plugins_dir.iterdir():
        if not entry.is_dir():
            continue
        manifest = entry / "elizaos.plugin.json"
        if not manifest.exists():
            continue
        try:
            data = json.loads(manifest.read_text())
            name = data.get("id") or f"@elizaos/{entry.name}"
            description = data.get("description", "")
            repo = (
                (data.get("repository", "") or "")
                .replace("https://github.com/", "")
                .replace(".git", "")
            )
            if not repo:
                repo = f"elizaos/{entry.name}"
            plugin = RegistryPlugin(
                name=name,
                git_repo=repo,
                git_url=data.get("repository") or f"https://github.com/{repo}.git",
                description=description,
                homepage=data.get("homepage"),
                topics=data.get("keywords", []),
                stars=0,
                language="TypeScript",
                npm_package=name,
                npm_v1_version=data.get("version"),
                npm_v2_version=data.get("version"),
                supports_v1=True,
                supports_v2=True,
                kind=data.get("kind"),
            )
            plugins[name] = plugin
            logger.debug("[registry] Found local plugin: %s (%s)", name, entry.name)
        except Exception as exc:
            logger.warning("[registry] Failed to parse %s: %s", manifest, exc)

    if plugins:
        logger.info("[registry] Loaded %d local plugins from %s", len(plugins), plugins_dir)
    return plugins


async def load_registry() -> dict[str, RegistryPlugin]:
    """Load the plugin registry.  Tries ``generated-registry.json`` first,
    falls back to ``index.json``, then merges in local plugins.
    Results are cached in-memory for 1 hour.
    """
    global _registry_cache
    if (
        _registry_cache is not None
        and (time.time() - _registry_cache["timestamp"]) < _CACHE_DURATION
    ):
        return _registry_cache["plugins"]

    logger.info("[registry] Fetching from next@registry...")
    plugins: dict[str, RegistryPlugin] = {}

    try:
        plugins = await _fetch_generated_registry()
        logger.info("[registry] Loaded %d plugins (generated-registry.json)", len(plugins))
    except Exception as exc:
        logger.warning(
            "[registry] generated-registry.json unavailable: %s, falling back to index.json",
            exc,
        )
        try:
            plugins = await _fetch_index_registry()
            logger.info("[registry] Loaded %d plugins (index.json)", len(plugins))
        except Exception as exc2:
            logger.warning(
                "[registry] index.json also unavailable: %s, using local plugins only",
                exc2,
            )

    # Merge local plugins (override remote)
    local = _scan_local_plugins()
    plugins.update(local)

    _registry_cache = {"plugins": plugins, "timestamp": time.time()}
    return plugins


# ---------------------------------------------------------------------------
# Lookup
# ---------------------------------------------------------------------------


def _resolve_plugin(
    registry: dict[str, RegistryPlugin],
    name: str,
) -> RegistryPlugin | None:
    p = registry.get(name)
    if p:
        return p
    if not name.startswith("@"):
        p = registry.get(f"@elizaos/{name}")
        if p:
            return p
    bare = re.sub(r"^@[^/]+/", "", name)
    for key, value in registry.items():
        if key.endswith(f"/{bare}") or key == bare:
            return value
    return None


# ---------------------------------------------------------------------------
# Search scoring
# ---------------------------------------------------------------------------


def _compute_search_score(plugin: RegistryPlugin, query: str) -> int:
    lq = query.lower()
    terms = [t for t in lq.split() if len(t) > 1]
    ln = plugin.name.lower()
    ld = plugin.description.lower()

    score = 0
    if ln == lq or ln == f"@elizaos/{lq}":
        score += 100
    elif lq in ln:
        score += 50

    if lq in ld:
        score += 30
    for t in plugin.topics:
        if lq in t.lower():
            score += 25

    for term in terms:
        if term in ln:
            score += 15
        if term in ld:
            score += 10
        for t in plugin.topics:
            if term in t.lower():
                score += 8

    if score > 0:
        if plugin.stars > 100:
            score += 5
        if plugin.stars > 500:
            score += 5
        if plugin.stars > 1000:
            score += 5

    return score


# ---------------------------------------------------------------------------
# Service
# ---------------------------------------------------------------------------


class PluginRegistryService(Service):
    """Fetches, caches, and searches the elizaOS plugin registry."""

    service_type: ClassVar[str] = "plugin_registry"

    @property
    def capability_description(self) -> str:
        return "Fetches, caches, and searches the elizaOS plugin registry"

    @classmethod
    async def start(cls, runtime: IAgentRuntime) -> PluginRegistryService:
        service = cls(runtime)
        logger.info("[PluginRegistryService] Started")
        return service

    async def stop(self) -> None:
        logger.info("[PluginRegistryService] Stopped")

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def get_registry_entry(self, name: str) -> RegistryPlugin | None:
        """Resolve a plugin by name with fuzzy matching."""
        registry = await load_registry()
        return _resolve_plugin(registry, name)

    async def search_plugins(
        self,
        query: str,
        limit: int = 10,
    ) -> list[PluginSearchResult]:
        """Search the registry by content query, returning scored results."""
        registry = await load_registry()
        scored: list[tuple[RegistryPlugin, int]] = []
        for plugin in registry.values():
            s = _compute_search_score(plugin, query)
            if s > 0:
                scored.append((plugin, s))

        scored.sort(key=lambda x: (-x[1], -x[0].stars))
        max_score = scored[0][1] if scored else 1

        return [
            PluginSearchResult(
                name=p.name,
                description=p.description,
                score=s / max_score,
                tags=p.topics,
                version=p.npm_v2_version or p.npm_v1_version or p.npm_v0_version,
                npm_package=p.npm_package,
                repository=f"https://github.com/{p.git_repo}",
                stars=p.stars,
                supports={"v0": p.supports_v0, "v1": p.supports_v1, "v2": p.supports_v2},
            )
            for p, s in scored[:limit]
        ]

    async def get_plugin_details(self, name: str) -> PluginMetadata | None:
        """Get full metadata for a plugin by name."""
        registry = await load_registry()
        plugin = _resolve_plugin(registry, name)
        return _to_metadata(plugin) if plugin else None

    async def get_all_plugins(self) -> list[PluginMetadata]:
        """Get metadata for all plugins in the registry."""
        registry = await load_registry()
        return [_to_metadata(p) for p in registry.values()]

    async def refresh(self) -> dict[str, RegistryPlugin]:
        """Force-refresh the registry cache."""
        reset_registry_cache()
        return await load_registry()
